src/Base/PyHandle.h

```diff
@@ -0,0 +1,152 @@
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef FREECAD_BASE_PYHANDLE_H
+#define FREECAD_BASE_PYHANDLE_H
+
+#include <CXX/Objects.hxx>
+#include "PyObjectBase.h"
+
+namespace Base {
+
+/** UnsafePyHandle: a convenient way to use python references as smart pointers for C++ objects.
+ * "Unsafe" stands for lack of runtime checks: type checking is limited; dereferencing the object
+ * if the C++ object was deleted will crash the application (you can check isValid() to guard
+ * against the latter).
+ *
+ * Usage example:
+ *     //create a py object, and save it into the handle
+ *     UnsafePyHandle<Base::Placement> h (new Base::PlacementPy(new Base::Placement), true);
+ *     //use the handle as if it is a pointer to C++ object
+ *     h->getPosition()
+ *     //and when the handle is destroyed, so is the py object, if it isn't referred to by something else.
+ *
+ * ## Memory management
+ *
+ * * when PyHandle is created, reference counter is incremented (except if
+ * "new_reference" constructor argument is true).
+ *
+ * * when PyHandle is destroyed, reference counter is decremented (no matter
+ * what "new_reference" value was passed to constructor). If resulting refcount
+ * is zero, the object is deleted.
+ *
+ * * Py::new_reference_to(PyHandle) increments reference count and returns PyObject*
+ *
+ * Cheatsheet:
+ *
+ * * when constructing from "new SomethingSomethingPy(new SomethingSomething)",
+ * use "new_reference". That is because SomethingSomethingPy sets reference counter
+ * to 1 in its constructor.
+ *
+ * * when constructing from getPyObject, use "new_reference"
+ *
+ * * when constructing from PyObject* you got as an argument of ...PyImp method
+ * implementation, don't use "new_reference".
+ *
+ * * when returning PyObject* from a getPyObject, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * * when returning PyObject* from a ...PyImp method implementation, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * The handle derives from PyCXX's Py::Object, so it can be conveniently used
+ * for adding to tuples and the like, when programming python API.
+ */
+ // Since it was made to become the main memory management construct of
+ // ConstraintSolver, checks for isValid() are intentionally not done to maximize
+ // performance. It is named "Unsafe" with the intention to have room for a safer
+ // "PyHandle" replica.
+template <class CppType>
+class UnsafePyHandle : public Py::Object
+{
+private:
```

Comment:
Another nitpick (so far your code looks great!) - I tend to prefer `public`, then `protected` then `private`, since the `public` is what is most interesting to a user, the `protected` is most interesting to a developer, and `private` should, for the most part, not interest anyone except whoever is extending or maintaining the class in question.

Reply:
makes sense